core/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosCoarseGrainedSchedulerBackend.scala

```diff
@@ -382,59 +382,97 @@ private[spark] class MesosCoarseGrainedSchedulerBackend(
+                None
+              } else {
+                Some((taskCPUs, taskMemory))
+              }
           }
-
-          tasks(offer.getId) ::= taskBuilder.build()
-          remainingResources(offerId) = resourcesLeft.asJava
-          totalCoresAcquired += taskCPUs
-          coresByTaskId(taskId) = taskCPUs
+        ) match {
+          case Some((taskCPUs: Int, taskMemory: Int)) =>
+            // Create a task
+            launchTasks = true
+            val taskId = newMesosTaskId()
+
+            slaves.getOrElseUpdate(slaveId, new Slave(offer.getHostname)).taskIDs.add(taskId)
+
+            val (afterCPUResources, cpuResourcesToUse) =
+              partitionResources(availableResources, "cpus", taskCPUs)
+            val (resourcesLeft, memResourcesToUse) =
+              partitionResources(afterCPUResources.asJava, "mem", taskMemory)
+
+            val taskBuilder = MesosTaskInfo.newBuilder()
+              .setTaskId(TaskID.newBuilder().setValue(taskId.toString).build())
+              .setSlaveId(offer.getSlaveId)
+              .setCommand(createCommand(offer, taskCPUs + extraCoresPerExecutor, taskId))
+              .setName("Task " + taskId)
+              .addAllResources(cpuResourcesToUse.asJava)
+              .addAllResources(memResourcesToUse.asJava)
+
+            sc.conf.getOption("spark.mesos.executor.docker.image").foreach { image =>
+              MesosSchedulerBackendUtil
+                .setupContainerBuilderDockerInfo(image, sc.conf, taskBuilder.getContainerBuilder)
+            }
+
+            tasks(offer.getId) ::= taskBuilder.build()
+            remainingResources(offerId) = resourcesLeft.asJava
+            totalCoresAcquired += taskCPUs
+            coresByTaskId(taskId) = taskCPUs
+          case None => logDebug(s"Insufficient offer CPU: $offerCpu  MEM: $offerMem")
         }
       }
     }
     tasks.toMap
   }
 
-  private def canLaunchTask(slaveId: String, resources: JList[Resource]): Boolean = {
-    val offerMem = getResource(resources, "mem")
-    val offerCPUs = getResource(resources, "cpus").toInt
-    val cpus = executorCores(offerCPUs)
-    val mem = executorMemory(sc)
-
-    cpus > 0 &&
-      cpus <= offerCPUs &&
-      cpus + totalCoresAcquired <= maxCores &&
-      mem <= offerMem &&
-      numExecutors() < executorLimit &&
-      slaves.get(slaveId).map(_.taskFailures).getOrElse(0) < MAX_SLAVE_FAILURES
+  /**
+   * Try and fit the resources to the constraints. Return None if it does not match
+   *
+   * @param sc Spark context
+   * @param availableCpus The available CPUs
+   * @param availableMem The available Memory
+   * @return Tuple of CPU (integer cores)  and Memory (integer MB) desired
+   */
+  private[mesos]
+  def calculateUsableResources(sc: SparkContext, availableCpus: Int, availableMem: Int):
+  Option[(Int, Int)] = {
+    val desiredMemory = executorMemory(sc)
+    val desiredCpu = executorCores(availableCpus)
```

Comment:
I like the idea, but it does make it harder to read. Can we refactor so we can collapse global and resource limits, and introduce like a case class for each condition, where it knows how to verify and logs error message when it doesn't match? 
I think it makes adding and understanding how offers are matched better.